src/extension_commands.ml

```diff
@@ -534,6 +534,193 @@ module Copy_type_under_cursor = struct
     command Extension_consts.Commands.copy_type_under_cursor handler
 end
 
+module Search_by_type = struct
+  let extension_name = "Search By Type"
+
+  let ocaml_lsp_doesnt_support_search_by_type ocaml_lsp =
+    not (Ocaml_lsp.can_handle_search_by_type ocaml_lsp)
+
+  let get_query_input ?previous_query ?(empty_result = false) () =
+    Promise.make @@ fun ~resolve ~reject:_ ->
+    let input_box =
+      let validationMessage =
+        Option.map previous_query ~f:(fun _ ->
+            InputBoxValidationMessage.create
+              ~message:
+                "No result found. Check the syntax or use a more general query."
+              ~severity:Warning
+              ())
+      in
+      InputBox.set
+        (Window.createInputBox ())
+        ~title:"Search By Type"
+        ~ignoreFocusOut:false
+        ?value:previous_query
+        ~placeholder:"int -> string / -int +string"
+        ?validationMessage:(if empty_result then validationMessage else None)
+        ~prompt:
+          "Perform a search by type request by providing a type signature to \
+           look for"
+        ()
+    in
+    let _disposable =
+      InputBox.onDidAccept
+        input_box
+        ~listener:(fun () ->
+          let query = InputBox.value input_box in
+          InputBox.set_busy input_box true;
+          InputBox.set_enabled input_box false;
+          resolve query)
+        ()
+    in
+    let _disposable =
+      InputBox.onDidChangeValue
+        input_box
+        ~listener:(fun _ -> InputBox.set_validationMessage input_box None)
+        ()
+    in
+    let _disposable =
+      InputBox.onDidHide input_box ~listener:(fun _ -> resolve None) ()
+    in
+    InputBox.show input_box
+
+  let get_search_results ~query ~limit ~with_doc ~position text_editor client =
+    let doc = TextEditor.document text_editor in
+    let uri = TextDocument.uri doc in
+    Custom_requests.(
+      send_request
+        client
+        Type_search.request
+        (Type_search.make ~uri ~position ~limit ~query ~with_doc ()))
+
+  let rec display_search_results query results text_editor position client =
+    let quickPickItems =
+      List.map
+        results
+        ~f:(fun (res : Custom_requests.Type_search.type_search_result) ->
+          QuickPickItem.create
+            ~label:res.name
+            ~description:res.typ
+            ~detail:(Option.value ~default:"" res.doc)
+            ())
+    in
+    let module QuickPick = Vscode.QuickPick.Make (QuickPickItem) in
+    let quickPick =
+      QuickPick.set
+        (Window.createQuickPick (module QuickPickItem) ())
```

Comment:
Do we really need functors and first class modules to make a binding for `QuickPick` and `createQuickPick` ?  I suspect there is a simpler approach.

Reply:
So the api indicates that we need to give it a type when invoking it (`QuickPick`)
https://code.visualstudio.com/api/references/vscode-api#QuickPick&lt;T&gt;. 
Looking through the codebase functors are what was used to write the bindings